src/cryptenroll/cryptenroll-tpm2.c

```diff
@@ -58,11 +60,77 @@ static int search_policy_hash(
         return -ENOENT; /* Not found */
 }
 
+static int get_pin(char **ret_pin_str, systemd_tpm2_flags *ret_flags) {
+        char *pin_str = NULL;
+        int r = 0;
+        systemd_tpm2_flags flags = 0;
+        _cleanup_strv_free_erase_ char **pin = NULL, **pin2 = NULL;
+
+        r = getenv_steal_erase("NEWPIN", &pin_str);
+        if (r < 0)
+                return log_error_errno(r, "Failed to acquire PIN from environment: %m");
+        if (r > 0) {
+                flags |= TPM2_FLAGS_USE_PIN;
+        } else {
+                for (int i = 5;; i--) {
+                        if (i <= 0)
+                                return log_error_errno(
+                                                SYNTHETIC_ERRNO(ENOKEY), "Too many attempts, giving up.");
+
+                        pin = strv_free_erase(pin);
+                        r = ask_password_auto(
+                                        "Please enter TPM2 PIN:",
+                                        "drive-harddisk",
+                                        NULL,
+                                        "tpm2-pin",
+                                        "cryptenroll.tpm2-pin",
+                                        USEC_INFINITY,
+                                        0,
+                                        &pin);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to ask for user pin: %m");
+                        assert(strv_length(pin) == 1);
```

Comment:
return an error instead of asserting, given it's user input?

Reply:
I adapted this from other usages of `ask_password_auto`. As far as I can tell it should always return a strv of length 1, it's not really influenced by user input.

Reply:
ok, leave it then